js/userDatabase.js

```diff
@@ -0,0 +1,131 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+"use strict";
+
+const OBJECT_STORE_PREFS = "prefs";
+const PINNED_LINKS_PREF = "pinnedLinks";
+
+(function(exports) {
+  const DATABASE_VERSION = 1;
+  const DATABASE_NAME = "NewTabData";
+
+  const OBJECT_STORE_PREFS_KEY = "prefType";
+  const OBJECT_STORE_PREFS_VALUE = "data";
+
+  const OPEN_DATABASE_TRANSACTION_STRING = "Open NewTabData connection";
+  const LOAD_DATA_TRANSACTION_STRING = "Load data ";
+  const UPDATE_DATA_TRANSACTION_STRING = "Update data ";
+  const READ_WRITE_TRANSACTION_STRING = "readwrite";
+
+  const gUserDatabase = {
+    _database: null,
+
+    init(callback) {
+      return new Promise((resolve, reject) => {
+        let request = window.indexedDB.open(DATABASE_NAME, DATABASE_VERSION);
+
+        request.onerror = event => {
+          gUserDatabase._logError(event, OPEN_DATABASE_TRANSACTION_STRING, reject);
```

Comment:
Why not have the error inline?

Also, an error string like "cannot open an indexedDB connection" might be more helpful in this case